src/main/java/org/testng/internal/AbstractParallelWorker.java

```diff
@@ -0,0 +1,116 @@
+package org.testng.internal;
+
+import org.testng.ClassMethodMap;
+import org.testng.IClassListener;
+import org.testng.ITestContext;
+import org.testng.ITestNGMethod;
+import org.testng.collections.Lists;
+import org.testng.internal.annotations.IAnnotationFinder;
+import org.testng.internal.thread.graph.IWorker;
+import org.testng.xml.XmlSuite;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+
+public abstract class AbstractParallelWorker {
+
+    public static AbstractParallelWorker newWorker(XmlSuite.ParallelMode mode) {
+        if (XmlSuite.ParallelMode.INSTANCES.equals(mode)) {
+            return new InstanceBasedParallelParallelWorker();
+        }
+        return new ClassBasedParallelParallelWorker();
+    }
+
+    public abstract List<IWorker<ITestNGMethod>> createWorkers(Arguments arguments);
+
+    public static class Arguments {
+        private List<ITestNGMethod> methods;
+        private IInvoker invoker;
+        private ConfigurationGroupMethods configMethods;
+        private ClassMethodMap classMethodMap;
+        private List<IClassListener> listeners;
+        private ITestContext testContext;
+        private IAnnotationFinder finder;
+
+        private Arguments() {
+            //We have a builder. Defeat instantiation via constructors.
+        }
+
+        public List<ITestNGMethod> getMethods() {
+            return Collections.unmodifiableList(methods);
+        }
+
+        public IInvoker getInvoker() {
+            return invoker;
+        }
+
+        public ConfigurationGroupMethods getConfigMethods() {
+            return configMethods;
+        }
+
+        public ClassMethodMap getClassMethodMap() {
+            return classMethodMap;
+        }
+
+        List<IClassListener> getListeners() {
+            return Collections.unmodifiableList(listeners);
+        }
+
+        public ITestContext getTestContext() {
+            return testContext;
+        }
+
+        public IAnnotationFinder getFinder() {
+            return finder;
+        }
+
+        public static class Builder {
+            private Arguments instance;
+
+            public Builder() {
+                instance = new Arguments();
+            }
+
+            public Builder usingMethods(List<ITestNGMethod> methods) {
```

Comment:
A nitpick but I think `using` is overly verbose. For builders, I think it's fine to use straight property names:

    new Builder()
        .methods(...)
        .invoker(...)
        .build()

Reply:
Fixed.